test/spec/php_spec.rb

```diff
@@ -0,0 +1,201 @@
+require_relative "spec_helper"
+
+describe "A PHP application" do
+	def self.genmatrix(matrix, keys = nil)
+		product_hash(matrix.select {|k,v| !keys || keys.include?(k) })
+	end
+	
+	def self.gencmd(args)
+		args.compact.map { |k,v|
+			if k.is_a? Numeric
+				ret = v.shellescape # it's an argument, so we want the value only
+			else
+				ret = k.shellescape
+				unless !!v == v # check if boolean
+					ret.concat(" #{v.shellescape}") # --foobar flags have no values
+				end
+			end
+			ret
+		}.join(" ").strip
+	end
+	
+	# the matrix of options and arguments to test
+	# we will generate all possible combinations of these using a helper
+	# nil means omitted
+	# anything with "broken" in the value will be treated as expected to fail
+	matrices = {
+		"apache2" => {
+			0 => [
+				"heroku-php-apache2"
+			],
+			'-C' => [
+				nil,
+				"conf/apache2.server.include.conf",
+				"conf/apache2.server.include.dynamic.conf.php",
+				"conf/apache2.server.include.broken"
+			],
+			'-F' => [
+				nil,
+				"conf/fpm.include.conf",
+				"conf/fpm.include.dynamic.conf.php",
+				"conf/fpm.include.broken"
+			],
+			1 => [ # document root argument
+				nil,
+				"docroot/",
+				"brokendocroot/"
+			]
+		},
+		"nginx" => {
+			0 => [
+				"heroku-php-nginx"
+			],
+			'-C' => [
+				nil,
+				"conf/nginx.server.include.conf",
+				"conf/nginx.server.include.dynamic.conf.php",
+				"conf/nginx.server.include.broken"
+			],
+			'-F' => [
+				nil,
+				"conf/fpm.include.conf",
+				"conf/fpm.include.dynamic.conf.php",
+				"conf/fpm.include.broken"
+			],
+			1 => [ # document root argument
+				nil,
+				"docroot/",
+				"brokendocroot/"
+			]
+		}
+	}
+	
+	context "with just an index.php" do
```

Comment:
Can you keep these tests, and split out the matrix / combination testing into a separate PR?

Reply:
separate file and commit now